tools/serverpod_cli/lib/src/logger/logger.dart

```diff
@@ -0,0 +1,111 @@
+import 'package:serverpod_cli/src/logger/loggers/std_out_logger.dart';
+
+/// Serverpods internal logger interface.
+/// All logging output should go through this interface.
+/// The purpose is to simplify implementing and switching out concrete logger
+/// implementations.
+abstract class Logger {
+  /// Sets the log level for the logger
+  void setLogLevel(LogLevel level);
+
+  /// Display debug [message] to the user.
+  /// Commands should use this for information that is important for
+  /// debugging purposes.
+  void debug(
+    String message, {
+    RawPrint style,
+  });
+
+  /// Display a normal [message] to the user.
+  /// Command should use this as the standard communication channel for
+  /// success, progress or information messages.
+  void info(
+    String message, {
+    RawPrint style,
+  });
+
+  /// Display a warning [message] to the user.
+  /// Commands should use this if they have important but not critical
+  /// information for the user.
+  void warning(
+    String message, {
+    RawPrint style,
+  });
+
+  /// Display an error [message] to the user.
+  /// Commands should use this if they want to inform a user that an error
+  /// has occurred.
+  void error(
+    String message, {
+    StackTrace? stackTrace,
+    RawPrint style,
+  });
+
+  /// Returns a [Future] that completes once all logging is complete.
+  Future<void> flush();
+}
+
+enum LogLevel {
+  debug,
+  info,
+  warning,
+  error,
+  nothing,
+}
+
+enum PrettyPrintType {
+  normal,
+  hint,
+  success,
+  list,
+  command,
+}
+
+class RawPrint {
+  final bool newParagraph;
+
+  const RawPrint({
+    this.newParagraph = false,
+  });
+}
+
+class BoxPrint extends RawPrint {
+  final String? title;
+  const BoxPrint({
+    this.title,
+    bool newParagraph = true,
+  }) : super(newParagraph: newParagraph);
+}
+
+class PrettyPrint extends RawPrint {
```

Comment:
Naming isn't great as the word implies some sort of action. To be more analogue with Flutter/Dart conventions perhaps a better name would be something like `AbstractConsoleTextStyle`, `ConsoleTextStyle`, and `ConsoleBoxTextStyle`.

Reply:
+1 one changing the names, but is Console the best word? Different loggers could potentially print to different streams, not necessarily a console. 

What about
`PrintStyle`, `BoxPrintStyle`, `FormattedPrintStyle`?